src/vcu/src/controls.c

```diff
@@ -0,0 +1,191 @@
+    // We shouldn't be braking in launch control, so don't worry about regen
+    set_brake_valve(false);
+    uint32_t front_wheel_speed = get_front_wheel_speed();
+
+    // Mini FSM
+    switch (lc_state) {
+      case BEFORE:
+        sendSpeedCmdMsg(0, 0);
+
+        // Transition
+        if (pedalbox_min(accel) > LC_ACCEL_BEGIN) {
+          lc_state = SPEEDING_UP;
+          printf("[LAUNCH CONTROL] SPEEDING UP STATE ENTERED\r\n");
+        }
+        break;
+      case SPEEDING_UP:
+        if (1000 > torque_command) sendSpeedCmdMsg(500, torque_command);
+        else sendSpeedCmdMsg(500, 1000);
+
+        // Transition
+        if (any_lc_faults()) {
+          lc_state = ZERO_TORQUE;
+          printf("[LAUNCH CONTROL] ZERO TORQUE STATE ENTERED\r\n");
+        } else if (front_wheel_speed > LC_WS_THRESH) {
+          lc_state = SPEED_CONTROLLER;
+          printf("[LAUNCH CONTROL] SPEED CONTROLLER STATE ENTERED\r\n");
+        }
+        break;
+      case SPEED_CONTROLLER:
+        speed_command = get_launch_control_speed(front_wheel_speed);
+        sendSpeedCmdMsg(speed_command, torque_command);
+
+        // Transition
+        if (any_lc_faults()) {
+          lc_state = ZERO_TORQUE;
+          printf("[LAUNCH CONTROL] ZERO TORQUE STATE ENTERED\r\n");
+        }
+      case ZERO_TORQUE:
+        sendTorqueCmdMsg(0);
+
+        if (pedalbox_max(accel) < LC_ACCEL_RELEASE) {
+          lc_state = DONE;
+          printf("[LAUNCH CONTROL] DONE STATE ENTERED\r\n");
+        }
+        break;
+    }
+
+  } else {
+    // Control regen brake valve:
+    bool brake_valve_state = control_settings.using_regen && get_pascals(pedalbox.REAR_BRAKE) < RG_REAR_BRAKE_THRESH;
+    set_brake_valve(brake_valve_state);
+
+    int32_t regen_torque;
+    if (brake_valve_state) {
+      regen_torque = get_regen_torque();
+    } else {
+      regen_torque = 0;
+    }
+
+    // Extra check to ensure we are only sending regen torque when allowed
+    if (torque_command == 0) {
+      torque_command = regen_torque;
+    }
+
+    sendTorqueCmdMsg(torque_command);
+  }
+}
+
+static int16_t get_torque(void) {
+  auto accel = pedalbox_avg(accel);
+
+  if (accel < PEDALBOX_ACCEL_RELEASE) return 0;
+
+  int16_t raw_torque =  MAX_TORQUE * (accel - PEDALBOX_ACCEL_RELEASE) / (MAX_ACCEL_VAL - PEDALBOX_ACCEL_RELEASE);
+  return raw_torque * control_settings.limp_factor / 100;
+}
+
+static int32_t get_regen_torque() {
+  int32_t regen_torque = 0;
+
+  // IMU velocity component unit conversion:
+  // Starting units: 100 * m/s, ending units: kph
+  // (100 m/s) * (1 km/1000 m) * (3600 s/1 hr) = 360 kph
+  // So multiply by 360 to get kph --> but this is squared, so multiply by 129600
+  int32_t imu_vel_north_squared = imu_velocity.north * imu_velocity.north;
+  int32_t imu_vel_east_squared = imu_velocity.east * imu_velocity.east;
+  int32_t car_speed_squared = 129600 * (imu_vel_north_squared + imu_vel_east_squared);
```

Comment:
Why are you squaring?

Reply:
Because I want the magnitude and have have components

Reply:
And I don't wanna take square root (is there a good way to do this?)

Reply:
Yeah I later do wrong math in the comparison